From 614a7f65fa2b53ea7d904dea20320a7a09b51700 Mon Sep 17 00:00:00 2001
From: Maciej Sobkowski <msobkowski@antmicro.com>
Date: Wed, 4 Jun 2025 13:15:24 +0200
Subject: [PATCH 12/16] subsys: instrumentation: rework UART handling

On MAX78002 EV KIT board if >1 character was transmitted (e.g. by running
`scripts/zaru.py`) RX was getting disabled and no further commands could be
sent. The reworked code is based on `samples/drivers/uart/echo_bot`.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>
---
 subsys/instrumentation/transport/uart.c | 19 ++++++++-----------
 1 file changed, 8 insertions(+), 11 deletions(-)

diff --git a/subsys/instrumentation/transport/uart.c b/subsys/instrumentation/transport/uart.c
index 08ddef57e6b..af93002066b 100644
--- a/subsys/instrumentation/transport/uart.c
+++ b/subsys/instrumentation/transport/uart.c
@@ -72,25 +72,22 @@ void handle_cmd(char *cmd)
 }
 
 __no_instrumentation__
-static void uart_isr(const struct device *dev, void *user_data)
+static void uart_isr(const struct device *uart_dev, void *user_data)
 {
-	int ret;
 	uint8_t byte = 0;
 	static uint8_t cur;
 
 	ARG_UNUSED(user_data);
 
-	while (uart_irq_update(dev) && uart_irq_is_pending(dev)) {
-		if (!uart_irq_rx_ready(dev)) {
-			continue;
-		}
+	if (!uart_irq_update(uart_dev)) {
+		return;
+	}
 
-		ret = uart_fifo_read(dev, &byte, 1);
-		if (ret < 0) {
-			uart_irq_rx_disable(dev);
-			return;
-		}
+	if (!uart_irq_rx_ready(uart_dev)) {
+		return;
+	}
 
+	while (uart_fifo_read(uart_dev, &byte, 1) == 1) {
 		if (!isprint(byte)) {
 			if (byte == '\r') {
 				_cmd_buffer[cur] = '\0';
-- 
2.43.0

